### PR TITLE
refs #26741 - switch to fedora-minimal base image

### PR DESCRIPTION
**Pull request by ohadlevy**
Not sure if this matters a lot, but this reduces the base image size from about 300MB to 130MB

### CODE REVIEW

**Comment by ohadlevy** [comment]
```
REPOSITORY                             TAG                 IMAGE ID            CREATED             SIZE
quay.io/ohadlevy/foreman               fedora-minimal      87fe4c0996ae        10 hours ago        634 MB
quay.io/foreman/foreman                develop             31565c65b906        4 days ago          809 MB
```

**Comment by ohadlevy** [comment]
and as a follow up, if anyone is interested to know what takes out the size:

```
docker image history quay.io/ohadlevy/foreman:fedora-minimal --no-trunc --format  '{{json .}}' |jq '"\(.Size) \(.CreatedBy)"' |grep -v 0B 
```
| Size | CMD |
|---|---|
| 29B | 1 HOME=/home/foreman /bin/sh -c date -u > BUILD_TIME |
| 29B |1 HOME=/home/foreman /bin/sh -c echo gem '\"rdoc\"' > bundler.d/container.rb && echo gem '\"tzinfo-data\"' >> bundler.d/container.rb |
| 61.5MB | /bin/sh -c #(nop) COPY --chown=1001:1001dir:d51fe1f0885fd791d74c90146503542babd0d0ccc55bece8930c943d44fa61ec in /home/foreman/public |
| 159MB | /bin/sh -c #(nop) COPY --chown=1001:1001dir:2d37cee2266e31da633cf8a8bcf0db211ddfa408772709ec89b8046f7c9c39bd in /home/foreman/vendor/ruby |
| 15.4kB |/bin/sh -c #(nop) COPY --chown=1001:1001file:4c2a97b612ab0f05dced02c5a6a3fc753c3192fafdd93959d8eb1ec6e3f000bd in /home/foreman/Gemfile.lock |
| 159B | /bin/sh -c #(nop) COPY --chown=1001:1001file:80bbadc813c41bbea83438fed20f7eb6c6299c3d8de3fdbf612ee7fd59e6a9f0 in /home/foreman/.bundle/config |
| 236B | /bin/sh -c #(nop) COPY file:af63372e99c77b3a939c4eeaee296844437568dcc933c95475417b559bd8c0bc in /usr/bin/entrypoint.sh |
| 18.7MB| /bin/sh -c #(nop) COPY --chown=1001:1001dir:c363e3f488ce8e3bb0db285ab17dc448f922ecda379e3b4b115a92e85dc29dde in /home/foreman/ |
| 236B |3 HOME=/home/foreman NODEJS_VERSION=11 RUBY_VERSION=2.6 /bin/sh -c chmod +x /usr/bin/entrypoint.sh |
| 236B |/bin/sh -c #(nop) COPY file:a10e921729afcec3219e1351fe6f87ea957d63d24a016213d7782a890e6cc7bb in /usr/bin/ |
| 2.9kB |3 HOME=/home/foreman NODEJS_VERSION=11 RUBY_VERSION=2.6 /bin/sh -c groupadd -r foreman -f -g 1001 && useradd -u 1001 -r -g foreman -d $HOME -s /sbin/nologin     -c \"Foreman Application User\" foreman &&     chown -R 1001:1001 $HOME |
| 264MB |2 NODEJS_VERSION=11 RUBY_VERSION=2.6 /bin/sh -c echo -e \"[nodejs]\\nname=nodejs\\nstream=${NODEJS_VERSION}\\nprofiles=\\nstate=enabled\\n\" > /etc/dnf/modules.d/nodejs.module &&   echo -e \"[ruby]\\nname=ruby\\nstream=${RUBY_VERSION}\\nprofiles=\\nstate=enabled\\n\" > /etc/dnf/modules.d/ruby.module && microdnf install mysql-libs mariadb-connector-c postgresql-libs ruby{,gems} rubygem-{rake,bundler} npm nc hostname python2-numpy && microdnf clean all |
| 131MB | fedora-minimal |

**Review by timogoebel — DISMISSED** [review]

File: Dockerfile [hunk 1]
```diff
@@ -1,17 +1,15 @@
...
+FROM registry.fedoraproject.org/fedora-minimal:30 as base
+ARG RUBY_VERSION="2.6"
+ARG NODEJS_VERSION="11"
 
 RUN \
-  echo "tsflags=nodocs" >> /etc/dnf/dnf.conf && \
-#  dnf -y upgrade && \
-  dnf -y module install ${RUBY_MODULE} ${NODEJS_MODULE} && \
-  dnf -y install mysql-libs mariadb-connector-c postgresql-libs ruby{,gems} rubygem-{rake,bundler} nc hostname \
+  echo -e "[nodejs]\nname=nodejs\nstream=${NODEJS_VERSION}\nprofiles=\nstate=enabled\n" > /etc/dnf/modules.d/nodejs.module && \
```
Comment (ekohl) [inline Dockerfile:7]: Not that I know exactly, but isn't there a DNF command to do this? Normally it's `dnf module enable nodejs:${NODEJS_VERSION}` or even `dnf module install ...`. Does microdnf provide a similar feature?

Comment (ohadlevy) [reply Dockerfile:7]: sadly no, see https://fedoramagazine.org/building-smaller-container-images/

Comment (ohadlevy) [reply Dockerfile:7]: or the actual source code https://github.com/rpm-software-management/microdnf/tree/master/dnf/plugins

Comment (ekohl) [reply Dockerfile:7]: I expected as much. Guess we now know why it's micro ;)

Comment (ohadlevy) [reply Dockerfile:7]: @ekohl the sad part is that at the end, our image include ruby, node and python :sob:  so much for micro....

**Review by ekohl — DISMISSED** [review]
I wonder if the tzinfo-data package should be installed via RPMs but I won't block on that.

**Comment by ohadlevy** [comment]
> I wonder if the tzinfo-data package should be installed via RPMs but I won't block on that.

for the record, its installed by rpm (hard dependency of many packages) but actually removed from the end image, see https://bugzilla.redhat.com/show_bug.cgi?id=1611117

**Comment by ekohl** [comment]
That BZ might be a useful comment on the line where you add it to the local bundler config.

**Comment by ohadlevy** [comment]
> That BZ might be a useful comment on the line where you add it to the local bundler config.

updated.

**Comment by iNecas** [comment]
I wonder what we actually gain from this? Since this image is mainly used for development, I would rather keep slightly bigger image, with the convenience that I don't have to learn all the small details on using `microdnf` instead of `dnf` and not being able to use the `dnf module`.

**Comment by ohadlevy** [comment]
> I wonder what we actually gain from this? Since this image is mainly used for development, I would rather keep slightly bigger image, with the convenience that I don't have to learn all the small details on using `microdnf` instead of `dnf` and not being able to use the `dnf module`.

I would argue we are saving the internet by reduce a couple of hundreds MB from many builds and containers :)

having said that, microdnf is not actually the real benefit, but rather out of the box much small footprint of needed packages, im assuming we can install regular dnf (since we already have python there) and use it (I dont suggest it because i want to remove the python dependency rather to add more).
if you actually try to use it, it works just as well with fedora regular dnf and module support is fairly trivial to configuring a yum repo. I still see it as net positive (and compatible with ubi8-minimal).

**Comment by iNecas** [comment]
If the concern is amount of downloaded data, base it of `ubi8-ruby25` or `ubi8-nodejs10` might actually save more downloads, one would not be re-downloading the platform rpms on every build, if the image is already cached.

**Comment by ohadlevy** [comment]
> If the concern is amount of downloaded data, base it of `ubi8-ruby25` or `ubi8-nodejs10` might actually save more downloads, one would not be re-downloading the platform rpms on every build, if the image is already cached.

I didnt use ubi as it has only ruby25, also our base image requires python/node and ruby, so I'm not sure if I'm seeing a lot more value there.

lastly, what makes sense imho is to build a foreman-base, foreman-dev base containers that are built often (e.g. daily) and then this would speed up building the actual images. (e.g. for every PR), but I felt it was too much to do in one step (nor I'm sure that people would find it useful), just like i ignored my entire foreman kubernetes yaml files ;)

**Comment by iNecas** [comment]
Ok, letting others to approve this step: I'm not convinced moving to minimal base image is the right direction for the time being (as the image is mainly for developers use, if I understand the usecase), as it involves some inconveniences (such as `dnf` and it's additional commands such as `dnf module module`, or small things like `vi`). I know those are all small things, I'm just thinking about avoiding any unnecessary friction in usability of this.

I'm not saying is absolutely terrible, I would just like to see more people happy with this move

**Comment by iNecas** [comment]
Another though after some fresh air: I might misunderstood the primary reason why the images are being build in the first place: I considered them more as basis for developers to build new containers on top of it, and for those cases, I would still argue the non-minimal would be better.

For other cases, such as just trying out the latest build in pretend-production mode, I agree less MBs is better - I need to give it a full try once I'm not on a phone internet.

If the goal of this is to get more closely to production-like setups, would people accept if the Dockerfile could actually produce containers both with fedora-minimal with Ruby 2.6 and Node 11, as well as ubi8-minimal with slightly older version of the stack, that is on the other hand closer to the real production setup we currently have?

With that, we would build foreman:fedora and foreman:ubi8 images, that could be used for example to compare differences between current and next production versions?

**Comment by ohadlevy** [comment]
> Another though after some fresh air: I might misunderstood the primary reason why the images are being build in the first place: I considered them more as basis for developers to build new containers on top of it, and for those cases, I would still argue the non-minimal would be better.

that's a valid usage case for sure, TBH I'm not sure why (when rebuilding it) you would want to add vi (especially you!), after all, you are not suppose to threat it as a server you login into...


> For other cases, such as just trying out the latest build in pretend-production mode, I agree less MBs is better - I need to give it a full try once I'm not on a phone internet.

yeah, smaller size works better for phone based internet as well :-)



> If the goal of this is to get more closely to production-like setups, would people accept if the Dockerfile could actually produce containers both with fedora-minimal with Ruby 2.6 and Node 11, as well as ubi8-minimal with slightly older version of the stack, that is on the other hand closer to the real production setup we currently have?

I think this is missing the whole point, why would you need to support different matrix of builds, IMHO we should pick one, and everyone can use it regardless of their favorite OS, if redhat would like to productize this at some stage, then, I guess it make sense to switch to ubi8 and just use that.

the whole point of using a container is that you have something that will always work, and you control exactly how it gets upgraded (e.g. no one can / should change the container so you can easily upgrade from one version to the next).

but imho this is bigger then the scope of the PR, perhaps move it to discourse?

**Comment by iNecas** [comment]
For me, it goes back to being close to the production setup: that's why I asked about using ubi instead of fedora. My understanding is the only reason to use fedora is to have newer ruby and node, so the reason why I suggested two variants is having one that is closer to the production setups, while keeping the edge possibility.

Will take it to discourse.

**Review by iNecas — APPROVED** [review]
Tested and works well (well, hit some issues during trying, but not related to this very PR)

**Comment by iNecas** [comment]
> lastly, what makes sense imho is to build a foreman-base, foreman-dev base containers that are built often (e.g. daily) and then this would speed up building the actual images. (e.g. for every PR), but I felt it was too much to do in one step (nor I'm sure that people would find it useful)

Playing with the builds more, I would find a use for `foreman-base` that would provide the base build environment (including things like node_modules), of which the `foreman` production image would be build. This way, rebuilding the dev images would be less of a hassle than it is today. So :+1: from me.

**Comment by iNecas** [comment]
@ohadlevy how about something like this https://github.com/theforeman/foreman/pull/6825 ?

**Pull request merged**